cubi_tk/irods_common.py

```diff
@@ -0,0 +1,250 @@
+        ).lower().startswith("y"):
+            self._save_irods_token(token)  # pragma: no cover
+        elif not self.ask:
+            self._save_irods_token(token)
+
+    def _save_irods_token(self, token: str):
+        """Retrieve PAM temp auth token 'obfuscate' it and save to disk."""
+        irods_auth_path = self.irods_env_path.parent.joinpath(".irodsA")
+        irods_auth_path.parent.mkdir(parents=True, exist_ok=True)
+
+        if isinstance(token, list) and token:
+            irods_auth_path.write_text(encode(token[0]))
+            irods_auth_path.chmod(0o600)
+        else:
+            logger.warning("No token found to be saved.")
+
+    @property
+    def session(self):
+        return self._init_irods()
+
+
+class iRODSTransfer(iRODSCommon):
+    """
+    Transfer files to iRODS.
+
+    :param jobs: Iterable of TransferJob objects
+    :type jobs: Union[list,tuple,dict,set]
+    """
+
+    def __init__(self, jobs: Iterable[TransferJob], **kwargs):
+        super().__init__(**kwargs)
+        self.__jobs = jobs
+        self.__total_bytes = sum([job.bytes for job in self.__jobs])
+        self.__destinations = [job.path_remote for job in self.__jobs]
+
+    @property
+    def jobs(self):
+        return self.__jobs
+
+    @property
+    def size(self):
+        return self.__total_bytes
+
+    @property
+    def destinations(self):
+        return self.__destinations
+
+    def _create_collections(self, job: TransferJob):
+        collection = str(Path(job.path_remote).parent)
+        with self.session as session:
+            session.collections.create(collection)
+
+    def put(self, recursive: bool = False, sync: bool = False):
+        # Double tqdm for currently transferred file info
+        # TODO: add more parenthesis after python 3.10
+        with tqdm(
+            total=self.__total_bytes,
+            unit="B",
+            unit_scale=True,
+            unit_divisor=1024,
+            position=1,
+        ) as t, tqdm(total=0, position=0, bar_format="{desc}", leave=False) as file_log:
+            for n, job in enumerate(self.__jobs):
+                file_log.set_description_str(
+                    f"File [{n + 1}/{len(self.__jobs)}]: {Path(job.path_local).name}"
+                )
+                try:
+                    with self.session as session:
+                        if recursive:
+                            self._create_collections(job)
+                        if sync and session.data_objects.exists(job.path_remote):
+                            t.update(job.bytes)
+                            continue
+                        session.data_objects.put(job.path_local, job.path_remote)
+                        t.update(job.bytes)
+                except Exception as e:  # pragma: no cover
+                    logger.error(f"Problem during transfer of {job.path_local}")
+                    logger.error(self.get_irods_error(e))
+                    sys.exit(1)
+            t.clear()
```

Comment:
The put function could maybe always trigger a self.chksum() after (successful) completion?
(At least as an option)

Reply:
Technically part of #202. I would leave this as is in the iRODSTransfer class. Ingest functions can automate the checksumming process or put it behind a -K flag. I would argue it's not really critical any more since SODAR calculates server-side checksums on LZ validation.